File: Topics/BTree/BTSerializeandDeserialize.java
package Topics.BTree;

import Libs.TreeNode;

import java.util.*;

public class BTSerializeandDeserialize {
    private String SPLIT = ",";
    private String NULL = "X";

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        StringBuilder sb = new StringBuilder();
        helperSerialize(sb, root);
        return sb.toString();
    }

    private void helperSerialize(StringBuilder sb, TreeNode root) {
        if (root == null) {
            sb.append(NULL).append(SPLIT);
        } else {
            sb.append(root.val).append(SPLIT);
            helperSerialize(sb, root.left);
            helperSerialize(sb, root.right);
        }
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        Queue<String> q = new LinkedList<>();
        q.addAll(Arrays.asList(data.split(SPLIT)));
        return helperDeserialize(q);
    }

    private TreeNode helperDeserialize(Queue<String> q) {
        String head = q.poll();
        if (head.equals(NULL)) {
            return null;
        } else {
            TreeNode root = new TreeNode(Integer.valueOf(head));
            root.left = helperDeserialize(q);
            root.right = helperDeserialize(q);
            return root;
        }
    }
}

File: Topics/BTree/BTSerializeandDeserialize.cpp
class Codec {
public:
    string serialize(TreeNode* root) {
        string result;
        preorder(root, result);
        return result;
    }

    TreeNode* deserialize(string data) {
        queue<string> q;

        stringstream ss(data);
        string token;

        while (getline(ss, token, ',')) {
            q.push(token);
        }

        return build(q);
    }

private:
    void preorder(TreeNode* root, string& result) {
        if (!root) {
            result += "X,";
            return;
        }

        result += to_string(root->val) + ",";

        preorder(root->left, result);
        preorder(root->right, result);
    }

    TreeNode* build(queue<string>& q) {
        string cur = q.front();
        q.pop();

        if (cur == "X") {
            return nullptr;
        }

        TreeNode* root = new TreeNode(stoi(cur));

        root->left = build(q);
        root->right = build(q);

        return root;
    }
};
